Sources/OpenAPIRuntime/Interface/ErrorHandlingMiddleware.swift

```diff
@@ -0,0 +1,86 @@
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftOpenAPIGenerator project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import HTTPTypes
+
+/// Error Handling middleware that converts an error to a HTTP response
+///
+/// This is an opt-in middleware that adopters may choose to include to convert application errors
+/// to a HTTP Response
+///
+/// Inclusion of this ErrorHandling Middleware should be accompanied by confirming errors to
+/// HTTPResponseConvertible protocol. Only errors  confirming to HTTPResponseConvertible are converted to a HTTP response. Other errors are re-thrown from this middleware.
+///
+/// Example usage
+/// 1. Create an error type that conforms to HTTPResponseConvertible protocol
+/// ```swift
+/// extension MyAppError: HTTPResponseConvertible {
+///    var httpStatus: HTTPResponse.Status {
+///    switch self {
+///        case .invalidInputFormat:
+///            .badRequest
+///        case .authorizationError:
+///            .forbidden
+///        }
+///    }
+/// }
+/// ```
+///
+/// 2. Opt in to the error middleware while registering the handler
+///
+/// ```swift
+/// let handler = try await RequestHandler()
+/// try handler.registerHandlers(on: transport, middlewares: [ErrorHandlingMiddleware()])
+///
+
+public struct ErrorHandlingMiddleware: ServerMiddleware {
+    public func intercept(_ request: HTTPTypes.HTTPRequest,
+                          body: OpenAPIRuntime.HTTPBody?,
+                          metadata: OpenAPIRuntime.ServerRequestMetadata,
+                          operationID: String,
+                          next: @Sendable (HTTPTypes.HTTPRequest, OpenAPIRuntime.HTTPBody?, OpenAPIRuntime.ServerRequestMetadata) async throws -> (HTTPTypes.HTTPResponse, OpenAPIRuntime.HTTPBody?)) async throws -> (HTTPTypes.HTTPResponse, OpenAPIRuntime.HTTPBody?) {
+        do {
+            return try await next(request, body, metadata)
+        } catch let error as ServerError {
+            if let appError = error.underlyingError as? (any HTTPResponseConvertible) {
+                return (HTTPResponse(status: appError.httpStatus, headerFields: appError.httpHeaderFields),
+                appError.httpBody)
+            } else {
+                throw error
+            }
+        }
+    }
+}
+
+/// Protocol used by ErrorHandling middleware to map an error to a HTTPResponse
+///
+/// Adopters who wish to convert their application error to a HTTPResponse
+/// should confirm their error(s) to this protocol
+
+public protocol HTTPResponseConvertible {
+    /// HTTP status to return in the response
+    var httpStatus: HTTPResponse.Status { get }
+    
+    /// (Optional) Headers to return in the response
+    var httpHeaderFields: HTTPTypes.HTTPFields { get }
+    
+    /// (Optional) The body of the response to return
+    var httpBody: OpenAPIRuntime.HTTPBody? { get }
+}
```

Comment:
I'll go and double check the proposal myself but I thought we were going to make these all optional? Right now, we have all three flavours:
- `httpStatus` is non-optional, with no default implementation;
- `httpHeaderFields` is non-optional, but has a default implementation; and
- `httpBody` is optional, with a default implementation.

IMO, if we're providing a default implementation in a protocol extension, we should _not_ use optional for the requirement.

I'll go double-check the proposal again, but, regardless of API spelling, is this the semantics we wanted here?

Specifically, if someone adds the `ErrorHandlingMiddleware`, don't we want it transform errors that do not conform into `500` and, in debug mode, possibly include something in the header or body?

Reply:
> IMO, if we're providing a default implementation in a protocol extension, we should not use optional for the requirement.

I'm not sure I agree, I think it's good the way it is - header fields is a container that can be empty, so it's non-optional and defaults to empty in the protocol extension. The body, however, is a stateful async sequence, and the middleware API allows returning `nil`, so I think it should be optional and default to `nil` in the protocol extension.

> Specifically, if someone adds the ErrorHandlingMiddleware, don't we want it transform errors that do not conform into 500 and, in debug mode, possibly include something in the header or body?

Yes I also caught this one, I believe [here](https://forums.swift.org/t/proposal-soar-0011-improved-error-handling/74736/28) we agreed to always return a response, and use 500 for unconformed errors.

Reply:
> > IMO, if we're providing a default implementation in a protocol extension, we should not use optional for the requirement.
> 
> I'm not sure I agree, I think it's good the way it is - header fields is a container that can be empty, so it's non-optional and defaults to empty in the protocol extension. The body, however, is a stateful async sequence, and the middleware API allows returning `nil`, so I think it should be optional and default to `nil` in the protocol extension.

Right I think the way it's currently structured and documented is conflating two things. The optionality of whether a user needs to provide an override, and the optionality of the underlying field. We should have these match the target response type in terms of optionality, which, as you point out, they do already. Concretely, the `HTTPResponse` has non-optional status and headers (although headers may be empty), and an _optional_ body.

I think what's confusing then is the use of "(Optional)" in the documentation comments for the protocol requirements. How about we remove that term from the documentation and provide defaults for everything in the protocol extension to make it clear, including `httpBody`, which will default to `nil` and make it clear what the default is in the protocol level documentation?

Reply:
You are right. I will convert undocumented errors to a 500 response. 
What do we want to do with header and body fields? Should we preserve the additional information in `ServerError.underlyingError` in the body as an application/json?

Reply:
There's precedent for this being done in debug and not in release. And I'm personally in favour of having this behaviour. Not having it is quite an annoying developer experience IMO.

I'd also propose making it configurable and defaulting to off. But it should be a backwards compatible change so happy to defer it altogether for now.

Reply:
I'd prefer to leave it out of this PR, however I would be supportive of a subsequent PR adding it as an explicit option. I really dislike the behavior being controlled by debug vs release.

Reply:
OK, @gayathrisairam feel free to break this out into a new GitHub issue where we can flesh out how this would look.

Reply:
https://github.com/apple/swift-openapi-generator/issues/677